From 5bce3179b862dab92810ed8536b8b3161284465b Mon Sep 17 00:00:00 2001
From: Andy Grover <agrover@redhat.com>
Date: Fri, 5 Sep 2014 16:36:14 -0400
Subject: [target] Don't allow setting WC emulation if device doesn't support

Message-id: <1409934989-29941-17-git-send-email-agrover@redhat.com>
Patchwork-id: 91474
O-Subject: [RHEL7.1 PATCH e-stor 16/31] target: Don't allow setting WC emulation if device doesn't support
Bugzilla: 1129387
RH-Acked-by: Chris Leech <cleech@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

commit 07b8dae38b09bcfede7e726f172e39b5ce8390d9 upstream.

Just like for pSCSI, if the transport sets get_write_cache, then it is
not valid to enable write cache emulation for it. Return an error.

see https://bugzilla.redhat.com/show_bug.cgi?id=1082675

Reviewed-by: Chris Leech <cleech@redhat.com>
Signed-off-by: Andy Grover <agrover@redhat.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/target/target_core_device.c b/drivers/target/target_core_device.c
index 660b109..8032917 100644
--- a/drivers/target/target_core_device.c
+++ b/drivers/target/target_core_device.c
@@ -796,10 +796,10 @@ int se_dev_set_emulate_write_cache(struct se_device *dev, int flag)
   pr_err("emulate_write_cache not supported for pSCSI\n");
   return -EINVAL;
  }
- if (dev->transport->get_write_cache) {
-  pr_warn("emulate_write_cache cannot be changed when underlying"
-   " HW reports WriteCacheEnabled, ignoring request\n");
-  return 0;
+ if (flag &&
+     dev->transport->get_write_cache) {
+  pr_err("emulate_write_cache not supported for this device\n");
+  return -EINVAL;
  }
 
  dev->dev_attrib.emulate_write_cache = flag;
-- 
1.7.1